atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -269,7 +270,7 @@ private enum State {
     private final ExpectationsMetrics expectationsDataCollectionMetrics;
 
     protected volatile boolean hasReads;
-    protected volatile boolean hasReadsThatRequireImmutableTimestampLockValidationAtCommitRound;
+    protected Set<UUID> readsRequiringImmutableTimestampValidationAtCommitRound;
```

Comment:
field needs to be `final` for JMM safety
```suggestion
    protected final Set<UUID> readsRequiringImmutableTimestampValidationAtCommitRound;
```